File: test-suite/python/security/open_redirect_buggy.py
from django.http import HttpResponseRedirect
from flask import redirect, request
from starlette.responses import RedirectResponse


def flask_login_redirect():
    next_url = request.args.get("next", "/")
    return redirect(next_url)


def flask_cookie_redirect():
    target = request.cookies.get("continue")
    return redirect(location=target)


def django_login_redirect(request):
    return HttpResponseRedirect(request.GET.get("next", "/"))


async def starlette_redirect(request):
    target = request.query_params.get("return_to", "/")
    return RedirectResponse(target)

File: test-suite/csharp/security/OpenRedirectBuggy.cs
using Microsoft.AspNetCore.Http;
using Microsoft.AspNetCore.Mvc;
using Microsoft.Extensions.Primitives;

public sealed class OpenRedirectBuggy : Controller
{
    public IActionResult RedirectQuery(HttpRequest request)
    {
        var target = request.Query["returnUrl"];
        return Redirect(target!);
    }

    public void ResponseRedirect(HttpContext context)
    {
        var next = context.Request.Query["next"];
        context.Response.Redirect(next!);
    }

    public IActionResult RedirectHeader(HttpContext context)
    {
        context.Request.Headers.TryGetValue("X-Return-To", out StringValues target);
        return new RedirectResult(target.ToString());
    }

    public void WriteLocationHeader(HttpContext context)
    {
        var destination = context.Request.Cookies["destination"];
        context.Response.Headers["Location"] = destination!;
    }

    public IActionResult RedirectRouteValue(HttpRequest request)
    {
        var callback = request.RouteValues["callback"]?.ToString();
        return RedirectPermanent(callback!);
    }
}
